### PR TITLE
MPC Edit Menu compilation error when compiling for multi-hotends

### PR DESCRIPTION
**Pull request by zeleps**
### Description

Compiling the latest bugfix with MPC_EDIT_MENU, multiple hotends, produces the following errors:

Marlin\src\lcd\menu\menu_advanced.cpp:355:13: error: 'c' is not captured
Marlin\src\lcd\menu\menu_advanced.cpp:369:61: error: 'mpc_edit_hotend' is not captured

Lambdas need byref capturing to use variables c / mpc_edit_hotend, but capturing lambdas have a different signature and are not convertible to screenFunc_t type. Converted mpc_edit_hotend to a static function and created a static index to the currently edited hotend in order to fix the issue.

I'm not very experienced with C++ lambdas, if there's another way to fix this, I'd love to see it.

### Related Items

PR #23984

### CODE REVIEW

**Comment by tombrazier** [comment]
This code is thinkyhead's improvement on what I wrote. I suspect mpc_edit_hotend should actually be a #define rather than a function. Menus are very clever but they still leave me a bit confused. What happens if you replace mpc_edit_hotend with 

```cpp
#define MPC_EDIT_HOTEND(N) \
  MPC_EDIT_DEFS(N); \
  START_MENU(); \
  BACK_ITEM(MSG_TEMPERATURE); \
  MPC_EDIT_ITEMS(N); \
  END_MENU();
```

and then use

```cpp
#define MPC_ENTRY(N) SUBMENU_N(N, MSG_MPC_EDIT, []{ MPC_EDIT_HOTEND(MenuItemBase::itemIndex); });
```

**Comment by zeleps** [comment]
I'm getting this now:

`Marlin\src\lcd\menu\menu_advanced.cpp:373:1: error: macro "SUBMENU_N_P" passed 4 arguments, but takes just 3`

It seems there is an issue with the expansion of macros START_MENU() and END_MENU() inside the lambda body. Still, when I remove these, I'm getting local variable not captured errors.

File: Marlin/src/lcd/menu/menu_advanced.cpp [hunk 1]
```diff
@@ -366,7 +366,7 @@ void menu_backlash();
           MPC_EDIT_ITEMS(e);
           END_MENU();
         };
-        #define MPC_ENTRY(N) SUBMENU_N(N, MSG_MPC_EDIT, []{ mpc_edit_hotend(MenuItemBase::itemIndex); });
+        #define MPC_ENTRY(N) SUBMENU_N(N, MSG_MPC_EDIT, [&]{ mpc_edit_hotend(MenuItemBase::itemIndex); });
```
Comment (thinkyhead) [inline Marlin/src/lcd/menu/menu_advanced.cpp:369]: The only change needed is to use the same default capture for the outer lambda as the inner lambda.

**Comment by thinkyhead** [comment]
> What happens if you replace `mpc_edit_hotend` with…

You end up with a distinct function for each hotend, instead of just one function that takes the hotend index.

File: Marlin/src/lcd/menu/menu_advanced.cpp (hunk 1 above)
Comment (zeleps) [reply Marlin/src/lcd/menu/menu_advanced.cpp:369]: I tried that first, along with line 354, doesn't compile:

`Marlin\src\lcd\menu\menu_item.h:275:40: error: cannot convert 'menu_advanced_temperature()::<lambda()>' to 'screenFunc_t' {aka 'void (*)()'}`

**Comment by zeleps** [comment]
@thinkyhead, this still doesn't compile:

```
Marlin\src\lcd\menu\menu_advanced.cpp:355:54: error: 'c' is not captured
Marlin\src\lcd\menu\menu_item.h:275:40: error: cannot convert 'menu_advanced_temperature()::<lambda()>' to 'screenFunc_t' {aka 'void (*)()'}
```

image

**Comment by tombrazier** [comment]
Confirmed. I get the same compile error with MPC_EDIT_MENU and HAS_MULTI_HOTEND.

**Comment by thisiskeithb** [comment]
@zeleps: You'll need to submit a new PR since this one is already merged.

**Comment by thinkyhead** [comment]
Missing capture? Usually you can just change `[]{ ... }` to `[&]{ ... }` and that solves it. But, that may not work for local autos.
```diff
- EDIT_ITEM_FAST_N(float43, N, MSG_MPC_AMBIENT_XFER_COEFF_FAN_E, &editable.decimal, 0, 1, []{ \
+ EDIT_ITEM_FAST_N(float43, N, MSG_MPC_AMBIENT_XFER_COEFF_FAN_E, &editable.decimal, 0, 1, [&]{ \
```

**Comment by zeleps** [comment]
> Missing capture? Usually you can just change `[]{ ... }` to `[&]{ ... }` and that solves it. But, that may not work for local autos.
> 
> ```diff
> - EDIT_ITEM_FAST_N(float43, N, MSG_MPC_AMBIENT_XFER_COEFF_FAN_E, &editable.decimal, 0, 1, []{ \
> + EDIT_ITEM_FAST_N(float43, N, MSG_MPC_AMBIENT_XFER_COEFF_FAN_E, &editable.decimal, 0, 1, [&]{ \
> ```

Of course I tried that (I mention it in the PR description), but it doesn't work:

```Marlin\src\lcd\menu\menu_item.h:275:40: error: cannot convert 'menu_advanced_temperature()::<lambda()>' to 'screenFunc_t' {aka 'void (*)()'}```

If I use [&] anywhere, the signature becomes incompatible with the ::action() method (explicitly casting it also fails). If I leave it as is, capture issues occur.

I did a lot of different tests before ending up with the static function as a last resort.

Configuration-wise, enable MPC edit menu and `#define HAS_MULTI_HOTEND 1` at some point before the function (e.g. @256). [This commit](https://github.com/zeleps/Marlin/commit/70f5f72bbbb7354046cd4f1f463c71c1f67e9439) demonstrates the problem.

**Pull request merged**